src/rust/engine/async_semaphore/src/lib.rs

```diff
@@ -0,0 +1,180 @@
+        .and_then(|permit| {
+          f().map(move |t| {
+            drop(permit);
+            t
+          })
+        }),
+    )
+  }
+}
+
+pub struct Permit {
+  inner: Arc<Mutex<Inner>>,
+}
+
+impl Drop for Permit {
+  fn drop(&mut self) {
+    let task = {
+      let mut inner = self.inner.lock().unwrap();
+      inner.available_permits += 1;
+      if let Some(task) = inner.waiters.pop_front() {
+        task
+      } else {
+        return;
+      }
+    };
+    task.notify();
+  }
+}
+
+pub struct PermitFuture {
+  inner: Option<Arc<Mutex<Inner>>>,
+}
+
+impl Future for PermitFuture {
+  type Item = Permit;
+  type Error = ();
+
+  fn poll(&mut self) -> Poll<Permit, ()> {
+    let inner = self.inner.take().expect("cannot poll PermitFuture twice");
+    let acquired = {
+      let mut inner = inner.lock().unwrap();
+      if inner.available_permits == 0 {
+        inner.waiters.push_back(task::current());
+        false
+      } else {
+        inner.available_permits -= 1;
+        true
+      }
+    };
+    if acquired {
+      Ok(Async::Ready(Permit { inner }))
+    } else {
+      self.inner = Some(inner);
+      Ok(Async::NotReady)
+    }
+  }
+}
+
+#[cfg(test)]
+mod tests {
+
+  use super::AsyncSemaphore;
+  use futures::{future, Future};
+  use std::sync::mpsc;
+  use std::thread;
+  use std::time::Duration;
+
+  #[test]
+  fn acquire_and_release() {
+    let sema = AsyncSemaphore::new(1);
+
+    sema
+      .with_acquired(|| future::ok::<_, ()>(()))
+      .wait()
+      .unwrap();
+  }
+
+  #[test]
+  fn at_most_n_acquisitions() {
+    let sema = AsyncSemaphore::new(1);
+    let handle1 = sema.clone();
+    let handle2 = sema.clone();
+
+    let (tx_thread1, acquired_thread1) = mpsc::channel();
+    let (unblock_thread1, rx_thread1) = mpsc::channel();
+    let (tx_thread2, acquired_thread2) = mpsc::channel();
+
+    thread::spawn(move || {
+      handle1
+        .with_acquired(move || {
+          // Indicate that we've acquired, and then wait to be signaled to exit.
+          tx_thread1.send(()).unwrap();
+          rx_thread1.recv().unwrap();
+          future::ok::<_, ()>(())
+        })
+        .wait()
+        .unwrap();
+    });
+
+    // Wait for thread1 to acquire, and then launch thread2.
+    acquired_thread1
+      .recv_timeout(Duration::from_secs(5))
+      .expect("thread1 didn't acquire.");
+
+    thread::spawn(move || {
+      handle2
+        .with_acquired(move || {
+          tx_thread2.send(()).unwrap();
+          future::ok::<_, ()>(())
+        })
+        .wait()
+        .unwrap();
+    });
+
+    // thread2 should not signal until we unblock thread1.
+    match acquired_thread2.recv_timeout(Duration::from_millis(500)) {
+      Err(_) => (),
+      Ok(_) => panic!("thread2 should not have acquired while thread1 was holding."),
+    }
+
+    // Unblock thread1 and confirm that thread2 acquires.
+    unblock_thread1.send(()).unwrap();
+    acquired_thread2
+      .recv_timeout(Duration::from_secs(5))
```

Comment:
This timeout should be the same as the recv timeout above - they're testing the same effect.

Ideally they'd both be more in the microsecond range - second-long blocking on an available semaphore isn't great, and even 500ms is pretty huge, but I guess there's a trade-off to be made for test reliability... Maybe 100ms is a reasonable compromise?

Reply:
The 5 second timeouts are what I refer to as "negative timeouts" (ie, we expect them _not_ to be reached), and the 500 millisecond timeout is a "positive timeout" because we expect it to be reached. Setting negative timeouts nice and high helps to reduce flakiness.

Will lower the positive timeout to 100ms though.

Reply:
Yeah, I get the thinking here, but I still think they should be the same. My reasoning goes:

If the test is flaky with the 5s timeout at 500ms, that means that sometimes it takes more than 500ms for us to wake up the thread.

Which means that waiting 500ms isn't waiting long enough to verify that we _won't_ wake up the thread, because we know that sometimes it takes longer than that.

Which means that our check which is trying to say "we never wake up the thread while the semaphore should be unavailable" needs to use a timeout bigger than 500ms to check that we never wake up the thread, because we _expect_ it to sometimes take more than 500ms (which is why the test is flaky).

So we need the values to be the same, because the amount we need to wait to convincingly guarantee that the thread _will_ be woken up is the same amount of time we need to wait to convincingly guarantee that the thread _won't_ be woken up.